src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -888,70 +1003,186 @@ scaleCreditsForMerge SingleRun{} _ = ScaledCredits 0
+supplyMergeCredits
+  (ScaledCredits c)
+  thresh@(CreditThreshold creditsThresh)
+  (MergingRun _ _ unspentCreditsVar mergeVar) = do
+    -- Add credits to the total of unspent credits
+    unspentCredits <- fetchAddInt unspentCreditsVar c
+    -- Compute our /estimate/ of what the new total of unspent credits is. The
+    -- /actual/ total might have been changed again by a different thread, but
+    -- that is okay because we will find out if\/when we run takeCredits.
+    let unspentCredits' = unspentCredits + c
+
+    -- Try to do some merging work if the new total now goes over the threshold.
+    --
+    -- TODO: when a merge is already completed, then could we skip most of the
+    -- stuff below?
+    when (unspentCredits' >= creditsThresh) $ do
+      isMergeDone <-
+        -- Try to take some unspent credits. The number of taken credits is the
+        -- number of merging steps we will try to do.
+        --
+        -- If an error happens during the body, then we put back as many credits
+        -- as we took, even if the merge has progressed. See Note [Credits] why
+        -- this is okay.
+        bracketOnError (tryTakeUnspentCredits unspentCreditsVar thresh (Credit unspentCredits'))
+                       (mapM_ (putBackUnspentCredits unspentCreditsVar)) $ \case
+          Nothing -> pure False
+          Just c' -> stepMerge mergeVar c'
+
+      -- If we just finished the merge, then we convert the output of the merge
+      -- into a new run. i.e., we complete the merge.
+      --
+      -- If an async exception happens before we get to perform the completion,
+      -- then that is fine. Either the next supplyMergeCredits or
+      -- expectCompletedMerge will complete the merge.
+      when isMergeDone $ completeMerge mergeVar
+
+{-# SPECIALISE tryTakeUnspentCredits ::
+     PrimVar RealWorld Int
+  -> CreditThreshold
+  -> Credit
+  -> IO (Maybe Credit) #-}
+-- | In a CAS-loop, subtract credits from the unspent credits to get it below
+-- the threshold again. If succesful, return Just that many credits, or Nothing
+-- otherwise.
+--
+-- The number of taken credits is a multiple of creditsThresh, so that the
+-- amount of merging work that we do each time is relatively uniform.
+--
+-- Nothing can be returned if the variable has already gone below the threshold,
+-- which may happen if another thread is concurrently doing the same loop on
+-- unspentCreditsVar.
+tryTakeUnspentCredits ::
+     PrimMonad m
+  => PrimVar (PrimState m) Int
+  -> CreditThreshold
+  -> Credit
+  -> m (Maybe Credit)
+tryTakeUnspentCredits !unspentCreditsVar thresh@(CreditThreshold !creditsThresh) (Credit !prev)
+  | prev < creditsThresh = pure Nothing
+  | otherwise = do
+      -- numThresholds is guaranteed to be >= 1
+      let !numThresholds = prev `div` creditsThresh
+      prev' <- casInt unspentCreditsVar prev (prev - numThresholds * creditsThresh)
+      if prev' == prev then
+        pure (Just (Credit prev))
```

Comment:
This seems dubious. We reduce the unspent credits by `numThreshold * creditsThresh`, but then return `prev`, which might be larger than that.
So we end up doing `prev` steps, which should bring the unspent credits to 0, but leave some residue in the counter?

Why do we do that whole "multiple of the threshold" thing in the first place? The two options I see are
1. using all accumulated credits
2. doing only a single threshold of work

How do multiples of the threshold give more uniform amounts of merging work than option 1? Is it because of the idea of using co-prime numbers, so it doesn't mess up those?

Option 2 gives a stronger upper bound on work done for a single `supplyMergeCredits`, but due to concurrent use might in rare cases leave more credits around than the threshold. This should usually quickly be fixed by the next updates, but it's something to be careful about.
(edit: It would be problematic with `supplyMergeCredits` being called with a huge number when opening a snapshot, but that could just call `stepMerge` directly as it doesn't have to deal with concurrent management of credits)

Reply:
> Calling supplyMergeCredits has the advantage of restoring the merge state and all the counters to exactly what it was when the snapshot was taken. We can probably do it with just stepMerge as well

I think `stepMerge` deals with everything but tracking the unspent credits, which arguable could just be set to 0. We are doing a big batch of merging already and don't have latency or concurrency constraints here. So why not spend all available credits? Seems like the simpler solution with less overhead.

Reply:
I think eventually we should store merge progress on disk when snapshotting, such that we don't have to redo the work. I've seen in the WP8 benchmark that it is rather slow to restore progress. The current implementation has close to the same effect as storing merge progress on disk, so I'd rather keep the impl like this

Reply:
> I've seen in the WP8 benchmark that it is rather slow to restore progress.

I think that's acceptable, but would be nice to avoid that work of course. The downside is that it can significantly increase the size of snapshots.

> The current implementation has close to the same effect as storing merge progress on disk, because the state is restored exactly the same, so I'd rather keep the impl like this.

Okay, makes sense. I can live with that.